File: javascript/complex-numbers/complex-numbers.js
/* eslint-disable  no-multi-spaces, indent */
/* eslint no-underscore-dangle: ["error", { "allowAfterThis": true }] */

export class ComplexNumber {
  constructor(a, b) {
    this.real = a;
    this.imag = b;
  }

  add(other) {
    if (!(other instanceof ComplexNumber)) throw new Error();
    return new ComplexNumber(this.real + other.real, this.imag + other.imag);
  }

  sub(other) {
    if (!(other instanceof ComplexNumber)) throw new Error();
    return new ComplexNumber(this.real - other.real, this.imag - other.imag);
  }

  mul(other) {
    if (!(other instanceof ComplexNumber)) throw new Error();
    const r = this.real * other.real - this.imag * other.imag;
    const i = this.imag * other.real + this.real * other.imag;
    return new ComplexNumber(r, i);
  }

  div(other) {
    if (!(other instanceof ComplexNumber)) throw new Error();
    const denom = ((other.real ** 2) + (other.imag ** 2));
    const r = (this.real * other.real + this.imag * other.imag) / denom;
    const i = (this.imag * other.real - this.real * other.imag) / denom;
    return new ComplexNumber(r, i);
  }

  get abs() {
    if (this._abs === undefined) {
      this._abs = Math.sqrt((this.real ** 2) + (this.imag ** 2));
    }
    return this._abs;
  }

  get conj() {
    if (this._conj === undefined) {
      this._conj = new ComplexNumber(this.real, -this.imag || 0); // avoid "-0"
    }
    return this._conj;
  }

  get exp() {
    if (this._exp === undefined) {
      this._exp = new ComplexNumber(Math.E ** this.real, 0)
                  .mul(new ComplexNumber(Math.cos(this.imag), Math.sin(this.imag)));
    }
    return this._exp;
  }
}

File: lua/complex-numbers/complex-numbers.lua
function math.hypot(x, y)  return math.sqrt(x^2 + y^2) end
function math.eq_ish(x, y) return math.abs(x - y) < 1e-6 end

local Complex, ComplexFunctions

Complex = function(r, i)
  local obj = {
    r = r,
    i = i or 0,
  }
  setmetatable(obj, ComplexFunctions)

  obj.abs  = function () return math.hypot(obj.r, obj.i) end
  obj.conj = function () return Complex(obj.r, -obj.i) end
  obj.exp  = function () 
    local ar = math.exp(obj.r)
    local br = math.cos(obj.i)
    local bi = math.sin(obj.i)
    return Complex(ar) * Complex(br, bi)
  end

  return obj
end

ComplexFunctions = {
  __eq = function (a, b)
    return math.eq_ish(a.r, b.r) and math.eq_ish(a.i, b.i)
  end,
  __add = function (a, b)
    local r = a.r + b.r
    local i = a.i + b.i
    return Complex(r, i)
  end,
  __sub = function (a, b)
    local r = a.r - b.r
    local i = a.i - b.i
    return Complex(r, i)
  end,
  __mul = function (a, b)
    local r = a.r * b.r - a.i * b.i
    local i = a.i * b.r + a.r * b.i
    return Complex(r, i)
  end,
  __div = function (a, b)
    local divisor = (b.r ^ 2 + b.i ^2)
    local r = (a.r * b.r + a.i * b.i) / divisor
    local i = (a.i * b.r - a.r * b.i) / divisor
    return Complex(r, i)
  end
}

return Complex
